trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/ACSLHandler.java

```diff
@@ -212,32 +222,80 @@ public Result visit(final IDispatcher main, final ACSLNode node) {
 	@Override
 	public Result visit(final IDispatcher main, final CodeAnnot node) {
 		if (node instanceof CodeAnnotStmt) {
-			final Check check;
-			if (mWitnessInvariantMode) {
-				check = new Check(Spec.WITNESS_INVARIANT);
-			} else {
-				check = new Check(Spec.ASSERT);
-			}
 			final ILocation loc = mLocationFactory.createACSLLocation(node);
-
-			ExpressionResult formula = (ExpressionResult) main
-					.dispatch(((Assertion) ((CodeAnnotStmt) node).getCodeStmt()).getFormula(), main.getAcslHook());
-
-			formula = mExprResultTransformer.transformSwitchRexIntToBool(formula, loc, main.getAcslHook());
-
 			final ExpressionResultBuilder resultBuilder = new ExpressionResultBuilder();
-			resultBuilder.addAllExceptLrValue(formula);
-
-			final AssertStatement assertStmt = new AssertStatement(loc, formula.getLrValue().getValue());
-			// TODO: Handle havoc statements
-			for (final Overapprox overapprItem : resultBuilder.getOverappr()) {
-				overapprItem.annotate(assertStmt);
+			final CodeStatement codeStmt = ((CodeAnnotStmt) node).getCodeStmt();
+			if (codeStmt instanceof Assertion) {
+				ExpressionResult formula =
+						(ExpressionResult) main.dispatch(((Assertion) codeStmt).getFormula(), main.getAcslHook());
+
+				formula = mExprResultTransformer.transformSwitchRexIntToBool(formula, loc, main.getAcslHook());
+
+				resultBuilder.addAllExceptLrValue(formula);
+
+				final AssertStatement assertStmt = new AssertStatement(loc, formula.getLrValue().getValue());
+				// TODO: Handle havoc statements
+				for (final Overapprox overapprItem : resultBuilder.getOverappr()) {
+					overapprItem.annotate(assertStmt);
+				}
+				resultBuilder.addStatement(assertStmt);
+				final List<HavocStatement> havocs = CTranslationUtil.createHavocsForAuxVars(formula.getAuxVars());
+				resultBuilder.addStatements(havocs);
+				final Check check;
+				if (mWitnessInvariantMode) {
+					check = new Check(Spec.WITNESS_INVARIANT);
+				} else {
+					check = new Check(Spec.ASSERT);
+				}
+				check.annotate(assertStmt);
+			}
+			if (codeStmt instanceof GhostUpdate) {
+				final GhostUpdate update = (GhostUpdate) codeStmt;
+				final ExpressionResult exprResult =
+						(ExpressionResult) main.dispatch(update.getExpr(), main.getAcslHook());

```

Comment:
What is the situation regarding (possibly implicit) assumptions or side effects resulting from the evaluation of the expression? For example:

- Wouldn't `// ghost x = 1/y`, where `y` is non-ghost, introduce some assumption that `y != 0`? Similarly for `*y` etc.
- What would happen for `// ghost x = y++` ?